### ISSUE TITLE
[upstream-repository.gradle] Auto-detect upstream repository URLs

### ISSUE THREAD

**Issue by kthoms** [eclipse/xtext#1222]
The `upstream-repository.gradle` scripts evaluate a property `upstreamBranch`. When not set, the branch name to use for upstream repositories is derived from the version string.

This approach is error prone and requires frequent source changes. When having changes spanning multiple repositories, a branch is usually created across several repositories with the same name. This is not necessarily the case for all upstream repositories.

Example: A change affects _xtext-extras_ and _xtext-eclipse_, but not xtext-lib. A branch "mychange" is created for _xtext-core_ and _xtext-extras_ then. When building _xtext-extras_, the upstream repository "mychange" has to be used for _xtext-core_, but for _xtext-lib_ the upstream repository should be taken for the "master" branch.

To successfully build either "upstreamBranch" property is set (e.g. by changing Jenkinsfile), or `upstream-repository.gradle` is modified. Both are code changes that have to be checked in and reverted before merge (while the last build with reverted URL might fail again).

The build should therefore be smart enough to select the right upstream repository URLs without any code change. Therefore  a repository's current branch name has to be evaluated and checked against Jenkins, if proper build jobs with the branch name exist for upstream repositories. If not, fallback to master should be performed.

The `adjustPipelines` script must not change `upstream-repository.gradle` files when working.

**Comment by cdietrich** [comment]
but then you need branches and built stuff on jenkins for all repos

**Comment by kthoms** [comment]
why? when each single repository could fall back to master? we can assume a master job always exists.

**Comment by cdietrich** [comment]
yes but the current logic does not support that.
and we need a solution for maven maven and  maven tycho too

**Comment by kthoms** [comment]
Right. One step after the other. This would not make adjust scripts obsolete for now, they would just do less. Ultimate goal is of course to get completely dynamic. I sense problems with the target platform resolution here.

**Comment by kthoms** [comment]
The given PRs are adding an autodetection to upstream-repositories.gradle. By default, upstream repositories are with function jenkinsPipelineRepo() and "branch" 'autodetect
```
maven { url jenkinsPipelineRepo('xtext-lib','autodetect') }
```

This will be used in the `upstream_branch` function to determine the current Git branch from the current repo. When set, also the tracking branch is considered, and finally 'master' as fallback.

With these branch names the Jenkins server is queried for an existing successful Maven repo for each upstream repo. If detected, the build will automatically choose the upstream build. If not, it will take 'master' from the upstream repo.

This behaviour can be overridden by either specifying the `upstreamBranch` property, or by setting an explicit branch name in the call to `jenkinsPipelineRepo()`. This is for example the case when adjustPipelines is executed. However, the script will become mostly obsolete with this, since it won't be necessary to change upstream-repositories.gradle anymore.

**Comment by cdietrich** [comment]
ok but we should not spend too much effort on such things now until we know where and how we will build in future

**Comment by kthoms** [comment]
Fixed for 2.16 M1
